### PR TITLE
Clean up distinction between package reference types

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

We've been handwaving along with the illusion that a `StoragePackageReference` was effectively a `LocalPackageReference` and could be treated as such for all intents and purposes ... this illusion broke down as I was implementing object renaming.

Upon inspection, it became clear that most (if not all) use of `AsLocalPackageReference` was interested in getting the `Group`, `Version` or both values from the reference.

By adding `GroupVersion()` to the interface definition, I was able to replace almost all calls to `AsLocalPackageReference` with cleaner code.

The few remaining references were easily refactored, allowing the removal of both the interface member `AsLocalPackageReference` and the helper method `PackageAsLocalPackage`.

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: v2/tools/generator/internal/astmodel/kubernetes_admissions_defaulter.go [hunk 1]
```diff
@@ -50,14 +50,16 @@ func (d *DefaulterBuilder) AddDefault(f *resourceFunction) {
 // as well as helper functions that allow additional handcrafted defaults to be injected by
 // implementing the genruntime.Defaulter interface.
 func (d *DefaulterBuilder) ToInterfaceImplementation() *InterfaceImplementation {
-	lpr, ok := d.resourceName.PackageReference.AsLocalPackage()
+	group, version, ok := d.resourceName.PackageReference.GroupVersion()
 	if !ok {
-		panic(fmt.Sprintf("expected resource name %s to be a local package reference", d.resourceName.String()))
+		panic(fmt.Sprintf("expected resource name %s to not have an external package reference", d.resourceName))
```
Comment (matthchr) [inline v2/tools/generator/internal/astmodel/kubernetes_admissions_defaulter.go:55]: Wondering if we should rephrase this to:
`unexpected external package reference for resource name %s`?

File: v2/tools/generator/internal/astmodel/kubernetes_admissions_validator.go [hunk 2]
```diff
@@ -65,14 +65,16 @@ func (v *ValidatorBuilder) AddValidation(kind ValidationKind, f *resourceFunctio
 // as well as helper functions that allow additional handcrafted validations to be injected by
 // implementing the genruntime.Validator interface.
 func (v *ValidatorBuilder) ToInterfaceImplementation() *InterfaceImplementation {
-	lpr, ok := v.resourceName.PackageReference.AsLocalPackage()
+	group, version, ok := v.resourceName.PackageReference.GroupVersion()
 	if !ok {
-		panic(fmt.Sprintf("expected resource name %s to be a local package reference", v.resourceName.String()))
+		panic(fmt.Sprintf("expected resource name %s to not have an external package reference", v.resourceName))
```
Comment (matthchr) [inline v2/tools/generator/internal/astmodel/kubernetes_admissions_validator.go:70]: Wondering if we should rephrase this to:
`unexpected external package reference for resource name %s`?

File: v2/tools/generator/internal/astmodel/storage_package_reference.go [hunk 3]
```diff
@@ -15,34 +15,58 @@ const (
 )
 
 type StoragePackageReference struct {
-	LocalPackageReference
+	inner   LocalPackageReference
+	version string
```
Comment (matthchr) [inline v2/tools/generator/internal/astmodel/storage_package_reference.go:19]: What's the difference between this `version` and the `version` inside the `LocalPackageReference`?

ohh, this one has the storage suffix:
`version: local.version + StoragePackageSuffix`

Might be worth clarifying in a comment that `inner` is the customer facing package that this storage package is related to (or however you wanna phrase it).

**Review by matthchr — APPROVED** [review]

File: v2/tools/generator/internal/astmodel/kubernetes_admissions_defaulter.go (hunk 1 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/astmodel/kubernetes_admissions_defaulter.go:55]: I like that phrasing. Done.

File: v2/tools/generator/internal/astmodel/kubernetes_admissions_validator.go (hunk 2 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/astmodel/kubernetes_admissions_validator.go:70]: Done.

File: v2/tools/generator/internal/astmodel/storage_package_reference.go (hunk 3 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/astmodel/storage_package_reference.go:19]: Done.
